File: dart-wrapper/cardinal_sdk/android/src/main/kotlin/com/icure/cardinal_sdk/UtilsPlugin.kt
package com.icure.cardinal_sdk

import com.icure.cardinal.sdk.dart.utils.ApiScope
import com.icure.cardinal.sdk.dart.utils.ResourceUtils
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

object UtilsPlugin : MethodChannel.MethodCallHandler {

	@Suppress("UNCHECKED_CAST")
	override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
		if (
			!dispatch(
				methodName = call.method,
				parameters = (call.arguments as Map<String, String>?).orEmpty()
			) { success, errorCode, errorMessage, errorDetail ->
				if (errorCode != null)
					result.error(errorCode, errorMessage, errorDetail)
				else
					result.success(success)
			}
		) result.notImplemented()
	}

	fun dispatch(
		methodName: String,
		parameters: Map<String, String>,
		resultCallback: (
			String?,
			String?,
			String?,
			String?,
		) -> Unit,
	): Boolean = when(methodName) {
		"releasePlatformResource" -> releasePlatformResource(parameters, resultCallback)
		"cancelJob" -> cancelJob(parameters, resultCallback)
		else -> null
	}?.let { true } ?: false

	private fun releasePlatformResource(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		ResourceUtils.releasePlatformResource(
			resultCallback,
			parameters.getValue("resourceId")
		)
	}

	private fun cancelJob(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		ApiScope.cancel(
			resultCallback,
			parameters.getValue("cancellationToken").toLong()
		)
	}
}

File: dart-wrapper/cardinal_sdk/ios/Classes/UtilsPlugin.swift
import Flutter
import CardinalDartSdkSupportLib

class UtilsPlugin {

	public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
		let args = call.arguments as! [String: String]
		let methodName = call.method

		let completed = UtilsPlugin.dispatch(methodName: methodName, parameters: args) { success, errorCode, errorMessage in
			if (errorCode != nil){
				result(FlutterError(code: errorCode!, message: errorMessage, details: nil))
			}
			else {
				result(success)
			}
		}

		if !completed {
			result(FlutterMethodNotImplemented)
		}
	}

	static func dispatch(
		methodName: String,
		parameters: [String : String],
		resultCallback: @escaping (
			String?,
			String?,
			String?
		) -> Void
	) -> Bool {
		switch methodName {
			case "releasePlatformResource": releasePlatformResource(parameters: parameters, resultCallback: resultCallback)
			case "cancelJob": cancelJob(parameters: parameters, resultCallback: resultCallback)
			default: return false
		}
		return true
	}

	private static func releasePlatformResource(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		ResourceUtils.shared.releasePlatformResource(
			dartResultCallback: resultCallback,
			resourceId: parameters["resourceId"]!
		)
	}
	
	private static func cancelJob(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		ApiScope.shared.cancel(
			dartResultCallback: resultCallback,
			cancellationToken: Int64(parameters["cancellationToken"]!),
		)
	}
}
